contracts/AirDrop.sol

```diff
@@ -0,0 +1,252 @@
+ */
+contract AirDrop is EIP712, IAirDrop, Ownable {
+    using SafeERC20 for IERC20;
+    using Address for address;
+
+    /**
+     * @dev Store balances of members
+     * @return amounts of available tokens & ether
+     */
+    mapping(address => uint256) public balanceOfTokens;
+    mapping(address => uint256) public balanceOfEther;
+
+    /**
+     * @dev Use for sign message with EIP712 standart
+     * @notice 'DROP_HASH' is signature of primaryType 'Drop'
+     */
+    bytes32 public constant DROP_HASH =
+        keccak256(
+            "Drop(uint256[] amount,uint256 deadline,address[] recepient,bool ether)"
+        );
+
+    IERC20 public token;
+
+    /**
+     * @dev Set 'token' IERC20 to interact with thrid party token
+     *
+     * Put address to state variable '_owner'
+     *
+     * @param token_ - of ERC20 contract
+     */
+    constructor(address token_) EIP712("AirDrop", "1") {
+        Address.isContract(token_);
+        token = IERC20(token_);
+    }
+
+    /**
+     * @dev send ERC20 tokens from 'owner_' balance to SC
+     *
+     * @param amount_ - how many tokens deposits
+     *
+     * NOTE : this function use wrapper of safeERC20 library
+     * check 'SafeERC20.sol' for more information
+     */
+    function depositTokens(uint256 amount_) external override onlyOwner {
+        require(amount_ != 0, "Error : 'Amount' , equal to 0");
+        token.safeTransferFrom(msg.sender, address(this), amount_);
+        emit DepositTokens(amount_);
+    }
+
+    /**
+     * @dev send ethers from 'owner_' balance to SC
+     *
+     * NOTE : if you send ether to 0 address , ethers will be blocked.
+     */
+    function depositEther() external payable override onlyOwner {
+        require(msg.value != 0, "Error : 'Amount' , equal to 0");
+        emit DepositEther(msg.value);
+    }
+
+    /**
+     * @dev make tokens available for members to withdraw
+     *
+     * @param recepients_ - address of member
+     * @param amount_ - how many tokens will be available
+     * @param deadline_ - how long signature will be valid
+     * @param v_ - can be only 27 or 28
+     * @param r_ - contain info to recover signer
+     * @param s_ - contain info to recover signer
+     *
+     * NOTE : more information 'EIP712'
+     */
+    function dropTokens(
+        address[] calldata recepients_,
+        uint256[] calldata amount_,
+        uint256 deadline_,
+        uint8 v_,
+        bytes32 r_,
+        bytes32 s_
+    ) external override onlyOwner {
+        require(
+            recepients_.length == amount_.length,
+            "Error : arrays have different length"
+        );
+        bytes32 structHash = keccak256(
+            abi.encode(
+                DROP_HASH,
+                keccak256(abi.encodePacked(amount_)),
+                deadline_,
+                keccak256(abi.encodePacked(recepients_)),
+                false
+            )
+        );
+        bytes32 digest = EIP712._hashTypedDataV4(structHash);
```

Comment:
you imitate EIP712, so it's simple call _hashTypedDataV4(structHash); without EIP712